Simplify README and prominently display recipes (#2349)
pytorch/torchtune: README.md

**Supervised Finetuning (SFT)**

| Type of Weight Update | 1 Device | >1 Device | >1 Node |
|-----------------------|:--------:|:---------:|:-------:|
| Full                  |    ✅    |     ✅    |   ✅    |
| [LoRA/QLoRA](https://pytorch.org/torchtune/stable/recipes/lora_finetune_single_device.html)            |    ✅    |     ✅    |    ❌    |

Example: ``tune run lora_finetune_single_device --config llama3_2/3B_lora_single_device`` <br />
You can also run e.g. ``tune ls lora_finetune_single_device`` for a full list of available configs.

**[Knowledge Distillation (KD)](https://pytorch.org/torchtune/0.4/tutorials/llama_kd_tutorial.html)**

| Type of Weight Update | 1 Device | >1 Device | >1 Node |
|-----------------------|:--------:|:---------:|:-------:|
| Full                  |    ❌    |     ❌    |    ❌    |
| LoRA/QLoRA            |    ✅    |     ✅    |    ❌    |

Example: ``tune run knowledge_distillation_distributed --config qwen2/1.5B_to_0.5B_KD_lora_distributed`` <br />
You can also run e.g. ``tune ls knowledge_distillation_distributed`` for a full list of available configs.

> **Reinforcement Learning + Reinforcement Learning from Human Feedback (RLHF)**

Reviewer: ```suggestion
**Preference Learning  / Reinforcement Learning**
```

Author: I agree this is the right way to think about it, but how many people would recognize Preference Learning over the term RLHF? I feel like RLHF is much more recognized for better or worse.

Reviewer: yeah it's your call

| Method | Type of Weight Update | 1 Device | >1 Device | >1 Node |
|------------------------------|-----------------------|:--------:|:---------:|:-------:|
| [DPO](https://pytorch.org/torchtune/stable/recipes/dpo.html)                          | Full                  |    ❌    |     ✅    |    ❌    |
|                           | LoRA/QLoRA            |    ✅    |     ✅    |    ❌    |

> | PPO                          | Full                  |    ✅    |     ❌    |    ❌    |

Reviewer: ```suggestion
| PPO (RLHF)                          | Full                  |    ✅    |     ❌    |    ❌    |
```
up to you

Reviewer: our upcoming GRPO implementation != RLHF. GRPO is just another optimization algorithm, the R1-style recipe we're going to land uses verifiable (rather than human rewards), so it's just RL. You could use PPO with verifiable rewards too, and it also wouldn't be RLHF

Author: Ahh I see, clearly have not looked closely enough at our PPO recipe.

* *November 2024*: torchtune has released [v0.4.0](https://github.com/pytorch/torchtune/releases/tag/v0.4.0) which includes stable support for exciting features like activation offloading and multimodal QLoRA
* *November 2024*: torchtune has added [Gemma2](recipes/configs/gemma2) to its models!
* *October 2024*: torchtune added support for Qwen2.5 models - find the configs [here](recipes/configs/qwen2_5/)
* *September 2024*: torchtune has support for **Llama 3.2 11B Vision**, **Llama 3.2 3B**, and **Llama 3.2 1B** models! Try them out by following our installation instructions [here](#Installation), then run any of the text configs [here](recipes/configs/llama3_2) or vision configs [here](recipes/configs/llama3_2_vision).

&nbsp;

## Overview 📚

torchtune is a PyTorch library for easily authoring, post-training and experimenting with LLMs. It provides:

- Hackable training recipes for SFT, knowledge distillation, RL and RLHF, and quantization-aware training
- Simple PyTorch implementations of popular LLMs like Llama, Gemma, Mistral, Phi, Qwen, and more

> - OOTB best-in-class memory efficiency, performance improvements, and scaling with the latest PyTorch APIs

Reviewer: ```suggestion
- OOTB best-in-class memory efficiency, performance improvements, and scaling, utilizing the latest PyTorch APIs
```
![image](https://github.com/user-attachments/assets/9c75df1c-53bf-49d2-bb3d-3b6730789abe)

## Overview 📚

> torchtune is a PyTorch library for easily authoring, post-training and experimenting with LLMs. It provides:

Reviewer: ![image](https://github.com/user-attachments/assets/d595e114-29c4-4d4c-86ce-69b0c1ae20da)
```suggestion
torchtune is a PyTorch library for easily authoring, post-training, and experimenting with LLMs. It provides:
```

Author: Haha I love the Oxford comma don't call me out like that

- YAML configs for easily configuring training, evaluation, quantization or inference recipes
- Built-in support for many popular dataset formats and prompt templates

&nbsp;

### Post-training recipes

> torchtune supports different kinds of post-training. A successful post-trained model will likely utilize several of the below methods.

Reviewer: maybe we should also just link to the recipe overview in here https://pytorch.org/torchtune/main/recipes/recipes_overview.html

> torchtune supports [different kinds of post-training](https://pytorch.org/torchtune/main/recipes/recipes_overview.html). A successful post-trained model will likely utilize several of the below methods.

Reviewer 2: ```suggestion
torchtune supports [the entire post-training lifecycle](https://pytorch.org/torchtune/main/recipes/recipes_overview.html). A successful post-trained model will likely utilize several of the below methods.
```

**Supervised Finetuning (SFT)**

| Type of Weight Update | 1 Device | >1 Device | >1 Node |
|-----------------------|:--------:|:---------:|:-------:|
| Full                  |    ✅    |     ✅    |   ✅    |

> | [LoRA/QLoRA](https://pytorch.org/torchtune/stable/recipes/lora_finetune_single_device.html)            |    ✅    |     ✅    |    ❌    |

Reviewer 2: wow no love for DoRA

Author: Nope, basically the same as LoRA / QLoRA and we say we support it in our actual docs.

> Example: ``tune run lora_finetune_single_device --config llama3_2/3B_lora_single_device``

Reviewer 2: OK so one thing that I like a little less about this -- it would be nice to have a starting point for each type of finetuning, and this kinda loses that. I do understand that the current table is a bit noisy, but maybe we can at least include something about using `tune ls recipe_name` to show existing configs for each of these

Author: Added.

**Knowledge Distillation (KD)**

| Type of Weight Update | 1 Device | >1 Device | >1 Node |
|-----------------------|:--------:|:---------:|:-------:|
| Full                  |    ❌    |     ❌    |    ❌    |

> | LoRA/QLoRA            |    ✅    |     ✅    |    ❌    |

Reviewer 2: Related to my other comment -- it does make me a little sad that we don't have any links to these. We really need to get at least some minimal docs page for each of our recipes. Do we have an issue tracking that?

Author: We really need to do this, wow.

Reviewer: https://github.com/pytorch/torchtune/issues/1408

Example: ``tune run knowledge_distillation_distributed --config qwen2/1.5B_to_0.5B_KD_lora_distributed``

**Reinforcement Learning + Reinforcement Learning from Human Feedback (RLHF)**

| Method | Type of Weight Update | 1 Device | >1 Device | >1 Node |
|------------------------------|-----------------------|:--------:|:---------:|:-------:|

> | [DPO](https://pytorch.org/torchtune/stable/recipes/dpo.html)                          | Full                  |    ✅    |     ❌    |    ❌    |

Reviewer 3: `full_dpo_distributed.py` works for >1 Device

Reviewer 3: please sync with https://github.com/pytorch/torchtune/pull/2363/files

|                           | LoRA/QLoRA            |    ✅    |     ✅    |    ❌    |
| PPO                          | Full                  |    ✅    |     ❌    |    ❌    |
|                           | LoRA/QLoRA            |    ❌    |     ❌    |    ❌    |
| GRPO                         | Full                  |    🚧    |     🚧    |   🚧    |
|                           | LoRA/QLoRA            |    ❌    |     ❌    |    ❌    |

Example: ``tune run lora_dpo_single_device --config llama3_1/8B_dpo_single_device``

> **Quantize Aware Training (QAT)**

Reviewer 2: ```suggestion
**Quantization-Aware Training (QAT)**
```

| Type of Weight Update | 1 Device | >1 Device | >1 Node |
|-----------------------|:--------:|:---------:|:-------:|
| [Full](https://pytorch.org/torchtune/stable/recipes/qat_distributed.html)                  |    ❌    |     ✅    |    ❌    |
| LoRA/QLoRA            |    ❌    |     ✅    |    ❌    |

Example: ``tune run qat_distributed --config llama3_1/8B_qat_lora``

The above configs are just examples to get you started. The full list of recipes can be found [here](recipes/). If you'd like to work on one of the gaps you see, please submit a PR! If there's a entirely new post-training method you'd like to see implemented in torchtune, feel free to open an Issue.

&nbsp;

> ### Models

Reviewer 2: Personally I don't really like the decreased emphasis on models. My assumption is that most people are gonna want to see pretty quickly what we support, and this buries it a bit too much. (It's fine to move it down, but idk why we need to take it out of table format)

Reviewer 4: I'd vote for something in between. Showing each model family we support and pointing to the latest version, but not as verbose as the current table.

Author: I went with something in-between. Happy?

&nbsp;

### Post-training recipes

torchtune supports different kinds of post-training. A successful post-trained model will likely utilize several of the below methods.

**Supervised Finetuning (SFT)**

| Type of Weight Update | 1 Device | >1 Device | >1 Node |
|-----------------------|:--------:|:---------:|:-------:|
| Full                  |    ✅    |     ✅    |   ✅    |
| LoRA/QLoRA            |    ✅    |     ✅    |         |

Example: ``tune run lora_finetune_single_device --config llama3_2/3B_lora_single_device``

**Knowledge Distillation (KD)**

| Type of Weight Update | 1 Device | >1 Device | >1 Node |
|-----------------------|:--------:|:---------:|:-------:|
| Full                  |          |           |         |
| LoRA/QLoRA            |    ✅    |     ✅    |         |

Example: ``tune run knowledge_distillation_distributed --config qwen2/1.5B_to_0.5B_KD_lora_distributed``

**Reinforcement Learning from Human Feedback (RLHF)**

| Method | Type of Weight Update | 1 Device | >1 Device | >1 Node |
|------------------------------|-----------------------|:--------:|:---------:|:-------:|

> | DPO                          | Full                  |    ✅    |           |         |

Reviewer: link to live docs (as and when we have them)? https://pytorch.org/torchtune/main/recipes/dpo.html

| [LoRA](https://pytorch.org/torchtune/main/tutorials/memory_optimizations.html#glossary-lora) | 8.5 | -51.61% | 8210 | +17.96% |
| [QLoRA](https://pytorch.org/torchtune/main/tutorials/memory_optimizations.html#quantized-low-rank-adaptation-qlora) | 4.6 | -45.71% | 8035 | -2.13% |

The final row in the table vs baseline + Packed Dataset uses **81.9%** less memory with a **284.3%** increase in tokens per second.

> <details>

Reviewer: This doesn't render correctly, in the rich diff at least

Author: Fixed.

&nbsp;

## Get Started 🚀

> To get started with torchtune, see our [First Finetune Tutorial](https://pytorch.org/torchtune/main/tutorials/first_finetune_tutorial.html). Our [End-to-End Workflow Tutorial](https://pytorch.org/torchtune/main/tutorials/e2e_flow.html) will show you how to evaluate, quantize and run inference with a Llama model. The rest of this section will provide a quick overview of these steps with Llama3.1.

Reviewer: ![image](https://github.com/user-attachments/assets/63b30481-4e0a-4d1c-a6c1-35a1e69688f8)

Author: Make it a PR suggestion so I can blindly accept.

> To get started with torchtune, see our [First Finetune Tutorial](https://pytorch.org/torchtune/main/tutorials/first_finetune_tutorial.html). Our [End-to-End Workflow Tutorial](https://pytorch.org/torchtune/main/tutorials/e2e_flow.html) will show you how to evaluate, quantize and run inference with a Llama model. The rest of this section will provide a quick overview of these steps with Llama3.1.

Reviewer: ```suggestion
To get started with torchtune, see our [First Finetune Tutorial](https://pytorch.org/torchtune/main/tutorials/first_finetune_tutorial.html). Our [End-to-End Workflow Tutorial](https://pytorch.org/torchtune/main/tutorials/e2e_flow.html) will show you how to evaluate, quantize, and run inference with a Llama model. The rest of this section will provide a quick overview of these steps with Llama3.1.
```
